File: src/backend/src/main/java/com/puneet/backend/entity/Monitor.java
package com.puneet.backend.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.UuidGenerator;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "monitors")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Monitor {

    @Id
    @GeneratedValue
    @UuidGenerator
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(nullable = false, length = 120)
    private String name;

    @Column(nullable = false, columnDefinition = "text")
    private String url;

    @Column(nullable = false, length = 10)
    @Builder.Default
    private String method = "GET";

    @Column(name = "expected_status_code", nullable = false)
    @Builder.Default
    private int expectedStatusCode = 200;

    @Column(name = "check_interval_sec", nullable = false)
    @Builder.Default
    private int checkIntervalSec = 60;

    @Column(name = "timeout_ms", nullable = false)
    @Builder.Default
    private int timeoutMs = 5000;

    @Column(name = "is_active", nullable = false)
    @Builder.Default
    private boolean active = true;

    @Column(name = "created_at", nullable = false)
    @Builder.Default
    private LocalDateTime createdAt = LocalDateTime.now();

    @Column(name = "updated_at", nullable = false)
    @Builder.Default
    private LocalDateTime updatedAt = LocalDateTime.now();

    @PreUpdate
    public void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Monitor m)) return false;
        return id != null && id.equals(m.id);
    }

    @Override
    public int hashCode() {
        return getClass().hashCode();
    }
}

File: src/frontend/src/types/monitor.ts
// Mirrors backend com.puneet.backend.dto.* for monitors.

export type HttpMethod = 'GET' | 'POST' | 'HEAD' | 'PUT';

export interface Monitor {
  id: string;
  name: string;
  url: string;
  method: HttpMethod;
  expectedStatusCode: number;
  checkIntervalSec: number;
  timeoutMs: number;
  active: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface CreateMonitorRequest {
  name: string;
  url: string;
  method?: HttpMethod;
  expectedStatusCode?: number;
  checkIntervalSec?: number;
  timeoutMs?: number;
}

// PATCH semantics: every field optional, only provided fields are applied.
export interface UpdateMonitorRequest {
  name?: string;
  url?: string;
  method?: HttpMethod;
  expectedStatusCode?: number;
  checkIntervalSec?: number;
  timeoutMs?: number;
  active?: boolean;
}
